src/main/java/gyro/google/compute/FirewallResource.java

```diff
@@ -0,0 +1,572 @@
+     *
+     * @subresource gyro.google.compute.FirewallAllowed
+     */
+    @Updatable
+    @ConflictsWith("denied")
+    public List<FirewallAllowed> getAllowed() {
+        if (allowed == null) {
+            allowed = new ArrayList<>();
+        }
+
+        return allowed;
+    }
+
+    public void setAllowed(List<FirewallAllowed> allowed) {
+        this.allowed = allowed;
+    }
+
+    /**
+     * A set of rules that deny requests to pass that get matched. Required if 'rule-type' set to ``DENY``.
+     *
+     * @subresource gyro.google.cloud.FirewallDenied
+     */
+    @Updatable
+    @ConflictsWith("allowed")
+    public List<FirewallDenied> getDenied() {
+        if (denied == null) {
+            denied = new ArrayList<>();
+        }
+
+        return denied;
+    }
+
+    public void setDenied(List<FirewallDenied> denied) {
+        this.denied = denied;
+    }
+
+    /**
+     * A set of destination IP in cidr form that the firewall rule applies to. Can only be set when 'direction' set to 'EGRESS'.
+     */
+    @Updatable
+    public Set<String> getDestinationRanges() {
+        if (destinationRanges == null) {
+            destinationRanges = new HashSet<>();
+        }
+
+        return destinationRanges;
+    }
+
+    public void setDestinationRanges(Set<String> destinationRanges) {
+        this.destinationRanges = destinationRanges;
+    }
+
+    /**
+     * The direction specifies the type of requests this rule applies to. INGRESS for incoming and EGRESS for outgoing requests. Valid values are ``INGRESS`` or ``EGRESS``. (Required)
+     */
+    @Required
+    @ValidStrings({"INGRESS","EGRESS"})
+    @Updatable
+    public String getDirection() {
+        return direction != null ? direction.toUpperCase() : null;
+    }
+
+    public void setDirection(String direction) {
+        this.direction = direction;
+    }
+
+    /**
+     * When true, disables the firewall rule. Defaults to ``false``.
+     */
+    @Updatable
+    public Boolean getDisabled() {
+        if (disabled == null) {
+            disabled = false;
+        }
+
+        return disabled;
+    }
+
+    public void setDisabled(Boolean disabled) {
+        this.disabled = disabled;
+    }
+
+    /**
+     * The priority of the firewall rule, when there are multiple that match a certain requests. Lower the number higher the priority. Defaults to ``1000``. Valid values are between ``0`` and ``65535``.
+     */
+    @Updatable
+    @Range(min = 0, max = 65535)
+    public Integer getPriority() {
+        if (priority == null) {
+            priority = 1000;
+        }
+
+        return priority;
+    }
+
+    public void setPriority(Integer priority) {
+        this.priority = priority;
+    }
+
+    /**
+     * A set of source IP in cidr form that the firewall rule applies to. Can only be set when 'direction' set to 'INGRESS'.
+     */
+    @Updatable
+    public Set<String> getSourceRanges() {
```

Comment:
This needs a value based validation https://github.com/perfectsense/gyro-google-provider/blob/feature/firewall-rules/src/main/java/gyro/google/compute/FirewallResource.java#L509-L511

In combination with https://github.com/perfectsense/gyro-google-provider/blob/feature/firewall-rules/src/main/java/gyro/google/compute/FirewallResource.java#L500-L506, it automatically avoids the conflicting scenario.

Reply:
I think you're arguing that your method of validation provides a better error message because other than that I don't see how my suggestion doesn't work.

How about this suggested changed to the `@Required` annotation to allow for a more specific error message?

```java
@Updatable
@Required(when = "direction", equals = "INGRESS", constraint = RequiredConstraint.ONLY_ONE)
public Set<String> getSourceRanges() {
...
}
```

The idea is that you'd annotate all fields that are potentially required when `direction` is set to `INGRESS`. The constraint determines if one field should be set, if all fields should be set, or if any field cannot be set.

Thoughts?